app/controllers/service/cypher_controller.rb

```diff
@@ -0,0 +1,24 @@
+# Cypher query via web services API
+
+class Service::CypherController < ServicesController
+
+  def index
+    authorize_user_from_token!
+    if params[:query]
+      # TBD: filter out create, delete, etc
+      cypher = params[:query]
+    else
+      # default for testing.  remove this before release.
+      cypher = "MATCH (n:Page) RETURN n.page_id LIMIT 10;"
+    end
+
+    # Web users are not supposed to modify the database.  The following
+    # is an ad hoc filter to try to prevent this, at least when it is 
+    # accidental; one shouldn't expect it to be secure against a concerted attack.
+    if cypher =~ /\w(delete|create|set|remove|merge|call|drop|load)\w/i
+      render status: 403    #403 = forbidden (401 means unauthenticated)
+    else
+      render json: TraitBank.query(cypher)
```

Comment:
I wonder, and this is purely speculative, but I do wonder... whether we should return the results of the call to #query, which is formatted more for neography (a gem) and includes some ancilary information, or if we should automatically (assuming the query was successful) return `TraitBank.query(cypher)['data']` (with some extra fluff in there to handle nil results, of course). Currently it will return something like this:

```
{"columns"=>["n.page_id"], "data"=>[[351759], [351760]]}
```

...And *perhaps* (not sure) all they want is the value of "data", here. I'm probably wrong. It might be worth expressing the nature of the return value in the docs somewhere, is all. [shrug] Ignore this comment, it's not actionable :D ...but I'm saving it for future discussion.

Reply:
My idea was to make the cypher service universal, implementing raw cypher, whose format someone can read about on the neo4j site. Also some scripts processing the output will need to know the column names, and it's easier to get them from "columns" than from the query itself. For general use we'll have a whole suite of specific queries for particular purposes, and the output will be reformatted to look more like json-ld (maybe even to _be_ json-ld), and I expect to phase out general use of the /cypher service.

Reply:
By the way if there's an error in parsing or otherwise processing the query, the diagnostic output should be reflected back in the 500 (or 400) response (currently just shows up on the server console). That's on my to-do list

Reply:
Oh I see, you're talking about the case where the query defaults? I should probably remove it, it's only there for debugging (so you don't have to always come up with a query when testing)